### PR TITLE
Add system tests for user lifecycle

### PR DESCRIPTION
**Pull request by bensheldon**
- Sets up Rails system tests, including capybara and capybara-email
- Creates a first system test that tests inviting a new user, confirmation, promotion to admin, demotion, and deletion
- ~~Within the development environment, replaces Sendmail delivery with letter_opener gem~~ unrelated to the system testing and removed
- Creates a binstub to start the job worker process during development

Note: I am using `rack_test` instead of Selenium Chromedriver because that has more complicated setup than I wanted to tackle right now.

This is a first step towards #34 by wrapping the existing auth system with tests.

### CODE REVIEW

File: bin/worker_dev [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+set -e
+
+QUEUE=* VERBOSE=1 RESQUE_PRE_SHUTDOWN_TIMEOUT=10 RESQUE_TERM_TIMEOUT=10 bundle exec rake environment resque:work
```
Comment (Mr0grog) [inline bin/worker_dev]: Nice, this is super useful. I always wind up consulting `Procfile` to remember what to run. 👍

File: test/system/user_test.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    Capybara.using_session(:admin) do
+      visit root_path
+
+      click_on "Login"
+
+      fill_in 'Email', with: admin.email
+      fill_in 'Password', with: 'testpassword'
+      click_on 'Log in'
+
+      assert page.has_content? "Logged in as #{admin.email}"
```
Comment (Mr0grog) [inline test/system/user_test.rb]: Would you please add a message to these assertions, so it’s a little clearer what’s happening if they fail? (Same with the other ones below.)

File: test/system/user_test.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
+require "application_system_test_case"
+
+class UsersTest < ApplicationSystemTestCase
```
Comment (Mr0grog) [inline test/system/user_test.rb]: Please make sure the file and class names match (the class is plural).

File: test/system/user_test.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      user_row = page.all('tr').find { |tr| tr.has_content? viewer_email }
+      within user_row do
+        click_on 'Promote to admin'
+      end
+    end
+
+    Capybara.using_session(:viewer) do
+      visit root_path
+
+      # TODO: Signing in again should not be necessary.
```
Comment (Mr0grog) [inline test/system/user_test.rb]: Oh, this is super weird! But I definitely cannot figure out what’s going wrong here either. :(

File: test/system/user_test.rb [hunk 5]
```diff
@@ -0,0 +1,103 @@
+require "application_system_test_case"
+
+class UsersTest < ApplicationSystemTestCase
+  test "Invite, register, confirm, promote, demote, delete" do
+    admin = users(:admin_user)
+    viewer_email = 'viewer@example.com'
+
+    Capybara.using_session(:admin) do
```
Comment (Mr0grog) [inline test/system/user_test.rb]: Would you mind adding a comment at the start of each of these blocks describing what it’s doing conceptually? This is a long function that flips between sessions, and could use description to make it easier to follow.

File: bin/worker_dev (hunk 1 above)
Comment (bensheldon) [reply bin/worker_dev]: 🎆

File: test/system/user_test.rb (hunk 2 above)
Comment (bensheldon) [reply test/system/user_test.rb]: I've added assertion messages to all `assert` and `refute` calls.

File: test/system/user_test.rb (hunk 3 above)
Comment (bensheldon) [reply test/system/user_test.rb]: Whoops! Renamed.

File: test/system/user_test.rb (hunk 4 above)
Comment (bensheldon) [reply test/system/user_test.rb]: Fixed it. The problem is that when clicking through the URL in the mailer, it was going to the `.dev` domain, whereas when doing `visit root_path` it was going to `127.0.0.1` and the cookies are not shared across domains, causing the user to appear to be logged out in the test.

I added additional test setup to set the default mailer host to be the Capybara host.

File: test/system/user_test.rb (hunk 5 above)
Comment (bensheldon) [reply test/system/user_test.rb]: Great idea! I added comments.

**Comment by bensheldon** [comment]
@Mr0grog I addressed your feedback and this is ready for re-review 🙌

File: test/system/users_test.rb [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+class UsersTest < ApplicationSystemTestCase
+  test "Invite, register, confirm, promote, demote, delete" do
+    admin = users(:admin_user)
+    viewer_email = 'user@example.com'
+
+    #
+    # Administrator sends a user invitation
+    #
+    Capybara.using_session('admin') do
```
Comment (Mr0grog) [inline test/system/users_test.rb]: Not an issue; just curious: why the switch from symbols to strings?

File: test/system/users_test.rb [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    end
+
+    #
+    # User is logged out from the deleted account and prevented from logging back in
+    #
+    Capybara.using_session('user') do
+      visit root_path
+      refute page.has_content?("Logged in as #{viewer_email}"), "User should NOT have an active session"
+
+      fill_in 'Email', with: viewer_email
```
Comment (Mr0grog) [inline test/system/users_test.rb]: This is failing for me now! I think this used to have `click_on 'Login'` before it.

On a separate note, I just realized that this test is not getting run in CI. Can you add another step for it in `.circleci/config.yml` after the existing test step? https://github.com/edgi-govdata-archiving/web-monitoring-db/blob/c2c3e57045c96b5a0adadd1e4d5b70dd08c98a73/.circleci/config.yml#L36-L41

I think the new step just needs to run `bundle exec rake test:system`, right?

Comment (bensheldon) [reply test/system/users_test.rb]: I went with `bin/rails test:system test` which is the magical order according to https://stackoverflow.com/a/49830043/241735

File: test/system/users_test.rb (hunk 6 above)
Comment (bensheldon) [reply test/system/users_test.rb]: That was a holdover from debugging the session logout problem. I changed it back to symbols.

**Comment by bensheldon** [comment]
@Mr0grog ok, I have the tests all passing. Let's give it another go 🏃

File: test/system/users_test.rb (hunk 7 above)
Comment (Mr0grog) [reply test/system/users_test.rb]: Huh, wonder what it is about Rake or the CLI parsing that makes the order so sensitive. Anyway, TIL!

**Pull request merged**